### ISSUE TITLE
Retrying failed jobs using `arq.Retry`

### ISSUE THREAD

**Issue by PerAStad** [samuelcolvin/arq#289]
Hi all!

We're using the branch in #259 with aioredis v2, and encountered some issues using the `arq.Retry` raising method ([link](https://arq-docs.helpmanual.io/#retrying-jobs-and-cancellation)) for scheduled cron jobs. @jonasks is currently checking if this is an issue in the current release, using aioredis v1.

## Use Case
- Update a token with a scheduled task from a third party application, ensuring my backend to always have a valid token at hand. 
- The third party application tends to respond with a gateway-timeout-error or something similar. Reason behind this can be many things, and can often times be fixed by simply performing another equal request, after a few seconds. 
  - I therefore wish to implement a sort of `cron job failed --> retrying job after X seconds` policy. 
  - By the looks of it, this seems to be a valid use case for the `arq.Retry` method. 

This is my `arq.py` file:
```python
import logging

from arq.cron import cron
from arq import Retry

from app.core.config import settings
from app.core.logging_config import setup_logging
import asyncio

log = logging.getLogger(__name__)


async def create_token(ctx: dict) -> None:
    """
    Create token and saving it to Redis
    """
    log.info('Creating token')
    # Pretending we fetching a token here:
    for attempt in range(2):
        if attempt != 1:
            # Take a little nap
            await asyncio.sleep(5)
            raise Retry(defer=ctx['job_try'] * 5)

    log.info('Token has been successfully created.')


async def startup(ctx: dict) -> None:
    """
    On startup. 
    """
    setup_logging()
    log.info('Arq worker is starting...')
    # await create_token(ctx)
    log.info('Arq worker startup was successful.')


class WorkerSettings:
    """
    ARQ Worker settings.
    """

    update_token_cron_job = cron(create_token, second={0, 30})

    cron_jobs = [update_token_cron_job]
    on_startup = startup

    redis_settings = settings.REDIS_SETTINGS

```
Running: `arq app.core.arq.WorkerSettings --watch app`:

Starts the arq worker correctly until the actual `update_token_cron_job` job is executed. This is the terminal output:
![image](https://user-images.githubusercontent.com/83281583/151167244-b2c520a3-b5fb-479f-9d1a-394adc455f55.png)

With the following error:
```
  File "/home/user/project/.venv/lib/python3.9/site-packages/aioredis/connection.py", line 868, in read_response
    raise response from None
aioredis.exceptions.ResponseError: Command # 3 (DEL) of pipeline caused error: ("wrong number of arguments for 'del' command",)
```

Also, if I include the job in the startup (uncommenting `await create_token(ctx)` under startup) sequence there is a keyError regarding `'job_try'`:
```
  File "/home/user/project/app/core/arq.py", line 23, in create_token
    raise Retry(defer=ctx['job_try'] * 5)
KeyError: 'job_try'
```

Examining the [docs](https://arq-docs.helpmanual.io/#retrying-jobs-and-cancellation), it can be seen that they are not using the `Retry` method for cron jobs but rather functions. Have I misunderstood how to use this `Retry` method, or is this a bug? All inputs are welcome!

Thank you for taking the time reading this :)

**Comment by JonasKs** [comment]
Can confirm, issue in current version, and not directly related to #259.

-----

`cat arq_example.py`:

<details>

```py

import asyncio

from arq import Retry
from arq.cron import cron
from arq.connections import RedisSettings


async def create_token(ctx: dict) -> None:
    """
    Create token and saving it to Redis
    """
    print('Creating token')
    # Pretending we fetching a token here:
    for attempt in range(2):
        if attempt != 1:
            # Take a little nap
            await asyncio.sleep(5)
            print('Will raise Retry-error.')
            raise Retry(defer=ctx['job_try'] * 5)

    print('Token has been successfully created.')


async def startup(ctx: dict) -> None:
    """
    On startup.
    """
    print('Arq worker is starting...')


class WorkerSettings:
    """
    ARQ Worker settings.
    """

    update_token_cron_job = cron(create_token, second={0, 30})

    cron_jobs = [update_token_cron_job]
    on_startup = startup
```
</details>

--------

`arq arq_example.WorkerSettings`

<details>

```bash
✗ arq arq_example.WorkerSettings

15:50:58: Starting worker for 1 functions: cron:create_token
15:50:58: redis_version=6.2.6 mem_usage=851.67K clients_connected=1 db_keys=0
Arq worker is starting...
15:51:00:   1.00s → cron:create_token()
Creating token
Will raise Retry-error.
15:51:05:   5.00s ↻ cron:create_token retrying job in 5.00s
Traceback (most recent call last):
  File "personal/arq_retry_test/.venv/bin/arq", line 8, in <module>
    sys.exit(cli())
  File "personal/arq_retry_test/.venv/lib/python3.9/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "personal/arq_retry_test/.venv/lib/python3.9/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "personal/arq_retry_test/.venv/lib/python3.9/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "personal/arq_retry_test/.venv/lib/python3.9/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "personal/arq_retry_test/.venv/lib/python3.9/site-packages/arq/cli.py", line 48, in cli
    run_worker(worker_settings_, **kwargs)
  File "personal/arq_retry_test/.venv/lib/python3.9/site-packages/arq/worker.py", line 756, in run_worker
    worker.run()
  File "personal/arq_retry_test/.venv/lib/python3.9/site-packages/arq/worker.py", line 256, in run
    self.loop.run_until_complete(self.main_task)
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "personal/arq_retry_test/.venv/lib/python3.9/site-packages/arq/worker.py", line 308, in main
    await self._poll_iteration()
  File "personal/arq_retry_test/.venv/lib/python3.9/site-packages/arq/worker.py", line 346, in _poll_iteration
    t.result()
  File "personal/arq_retry_test/.venv/lib/python3.9/site-packages/arq/worker.py", line 590, in run_job
    await asyncio.shield(
  File "personal/arq_retry_test/.venv/lib/python3.9/site-packages/arq/worker.py", line 626, in finish_job
    await tr.execute()
  File "personal/arq_retry_test/.venv/lib/python3.9/site-packages/aioredis/commands/transaction.py", line 176, in execute
    return await self._do_execute(
  File "personal/arq_retry_test/.venv/lib/python3.9/site-packages/aioredis/commands/transaction.py", line 296, in _do_execute
    return (await self._gather_result(return_exceptions))
  File "personal/arq_retry_test/.venv/lib/python3.9/site-packages/aioredis/commands/transaction.py", line 198, in _gather_result
    raise self.error_class(errors)
aioredis.errors.MultiExecError: ('MultiExecError errors:', [TypeError("delete() missing 1 required positional argument: 'key'")])

```

</details>


--------

`cat pyproject.toml`: 

<details> 


```toml
[tool.poetry]
name = "arq_retry_test"
version = "0.1.0"
description = ""
authors = ["Jonas <jonas-ks@hotmail.com>"]

[tool.poetry.dependencies]
python = "^3.9"
arq = "^0.22"

[tool.poetry.dev-dependencies]

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"

```

</details>

**Comment by JonasKs** [comment]
Seems like there is a test for this [here](https://github.com/samuelcolvin/arq/blob/fix-tests/tests/test_cron.py#L149-L163) (and in your new [fix-test](https://github.com/samuelcolvin/arq/blob/fix-tests/tests/test_cron.py#L149-L163) branch, @samuelcolvin ), so I have no idea why ours keep crashing. I tried raising the same exception here, but same error applies:


```
aioredis.errors.MultiExecError: ('MultiExecError errors:', [TypeError("delete() missing 1 required positional argument: 'key'")])
```
I've tried `redis:latest`, `redis:5.0.14`, `redis:6.2.6` and `redis:6.2.6-alpine`. Any idea?

**Comment by JonasKs** [comment]
Ah. Found the workaround, and it has already been implemented in #273.
